README.md
# simple-utils ![npm (scoped)](https://img.shields.io/npm/v/@jungleford/simple-utils.svg)

A set of simple utilities for internal usage.

## Install

```bash
$ npm install
```

## Usage

For ES6,

```js
import {Utils, Libs} from '@jungleford/simple-utils';
```

For ES5,

```js
var Utils = require('@jungleford/simple-utils').Utils;
var Libs = require('@jungleford/simple-utils').Libs;
```

### Utils

A set of utility methods.

### Libs

A wrapper of third-party libraries.

For some scenarios (e.g., WeiXin mini program), the libraries (like `assert`, or `_` in Lodash) couldn't be recognized. So here provides a default implementation with standard JavaScript.

```js
const assert = Libs.assert;
const _ = Libs._;
```

## Testing

You need a mocha or karma command line utility, for this package, mocha@5.2.0, mocha-webpack@2.0.0-beta.0 and karma@3.1.4 are recommended.

Inline script is available:

```bash
$ npm test
```

or

```bash
$ npm run karma
```

CHANGELOG.md
### [Versions](https://github.com/jungleford/simple-utils/releases)

## 0.1.0
* Wrap third-party libraries, in order to adapt non-nodejs environment (e.g. WeiXin mini program).

## 0.0.4
* Stable version. Separate this bundle from [math-folding UI](https://github.com/jungleford/math-folding-react).

test/js/utils.spec.js
import {expect} from 'chai';

import {Utils} from '../../src/js';

describe('[@jungleford/simple-utils] test suite:', () => {

  it('generateGradualColors()', () => {
    // Invalid arguments
    expect(Utils.generateGradualColors.bind(Utils, 0)).to.throw('`count` must be a positive integer.');
    expect(Utils.generateGradualColors.bind(Utils, -1)).to.throw('`count` must be a positive integer.');
    expect(Utils.generateGradualColors.bind(Utils, 3.5)).to.throw('`count` must be a positive integer.');
    expect(Utils.generateGradualColors.bind(Utils, '1')).to.throw('`count` must be a positive integer.');
    expect(Utils.generateGradualColors.bind(Utils, [1, 2])).to.throw('`count` must be a positive integer.');
    expect(Utils.generateGradualColors.bind(Utils, {number: 1})).to.throw('`count` must be a positive integer.');
    expect(Utils.generateGradualColors.bind(Utils, null)).to.throw('`count` must be a positive integer.');
    expect(Utils.generateGradualColors.bind(Utils, undefined)).to.throw('`count` must be a positive integer.');

    // Expected results
    expect(Utils.generateGradualColors(1)).to.have.lengthOf(1);
    expect(Utils.generateGradualColors(1)).to.eql(['#000000']); // deep equal
    expect(Utils.generateGradualColors(4)).to.have.lengthOf(4);
    expect(Utils.generateGradualColors(4)).to.eql(['#000000', '#400000', '#800000', '#C00000']); // deep equal
  });

  it('generateGradualColorMatrix()', () => {
    // Invalid arguments
    expect(Utils.generateGradualColorMatrix.bind(Utils, 0)).to.throw('`rowCount` must be a positive integer.');
    expect(Utils.generateGradualColorMatrix.bind(Utils, -1)).to.throw('`rowCount` must be a positive integer.');
    expect(Utils.generateGradualColorMatrix.bind(Utils, 3.5)).to.throw('`rowCount` must be a positive integer.');
    expect(Utils.generateGradualColorMatrix.bind(Utils, '1')).to.throw('`rowCount` must be a positive integer.');
    expect(Utils.generateGradualColorMatrix.bind(Utils, [1, 2])).to.throw('`rowCount` must be a positive integer.');
    expect(Utils.generateGradualColorMatrix.bind(Utils, {number: 1})).to.throw('`rowCount` must be a positive integer.');
    expect(Utils.generateGradualColorMatrix.bind(Utils, null)).to.throw('`rowCount` must be a positive integer.');
    expect(Utils.generateGradualColorMatrix.bind(Utils, undefined)).to.throw('`rowCount` must be a positive integer.');

    // Expected results
    expect(Utils.generateGradualColorMatrix(1)).to.have.lengthOf(1);
    expect(Utils.generateGradualColorMatrix(1)[0]).to.have.lengthOf(1);
    expect(Utils.generateGradualColorMatrix(1)).to.eql([
                                                         ['#000000']
                                                       ]); // deep equal
    let colors4x4 = Utils.generateGradualColorMatrix(4);
    expect(colors4x4).to.have.lengthOf(4);
    colors4x4.forEach(row => {
      expect(row).to.have.lengthOf(4);
    });
    expect(colors4x4).to.eql(
      [
        ['#000000', '#004000', '#008000', '#00C000'],
        ['#400000', '#404000', '#408000', '#40C000'],
        ['#800000', '#804000', '#808000', '#80C000'],
        ['#C00000', '#C04000', '#C08000', '#C0C000']
      ]
    ); // deep equal
  });

  it('getReverseColor()', () => {
    // Invalid arguments
    expect(Utils.getReverseColor.bind(Utils, 0)).to.throw('`color` must be a positive integer with hex format between #000000 and #FFFFFF.');
    expect(Utils.getReverseColor.bind(Utils, -1)).to.throw('`color` must be a positive integer with hex format between #000000 and #FFFFFF.');
    expect(Utils.getReverseColor.bind(Utils, 3.5)).to.throw('`color` must be a positive integer with hex format between #000000 and #FFFFFF.');
    expect(Utils.getReverseColor.bind(Utils, 1)).to.throw('`color` must be a positive integer with hex format between #000000 and #FFFFFF.');
    expect(Utils.getReverseColor.bind(Utils, 'X')).to.throw('`color` must be a positive integer with hex format between #000000 and #FFFFFF.');
    expect(Utils.getReverseColor.bind(Utils, '1111111')).to.throw('`color` must be a positive integer with hex format between #000000 and #FFFFFF.');
    expect(Utils.getReverseColor.bind(Utils, '@11111')).to.throw('`color` must be a positive integer with hex format between #000000 and #FFFFFF.');
    expect(Utils.getReverseColor.bind(Utils, '#1111111')).to.throw('`color` must be a positive integer with hex format between #000000 and #FFFFFF.');
    expect(Utils.getReverseColor.bind(Utils, 'abcdefg')).to.throw('`color` must be a positive integer with hex format between #000000 and #FFFFFF.');
    expect(Utils.getReverseColor.bind(Utils, '#abcdefg')).to.throw('`color` must be a positive integer with hex format between #000000 and #FFFFFF.');
    expect(Utils.getReverseColor.bind(Utils, 'ABCDEFG')).to.throw('`color` must be a positive integer with hex format between #000000 and #FFFFFF.');
    expect(Utils.getReverseColor.bind(Utils, '#ABCDEFG')).to.throw('`color` must be a positive integer with hex format between #000000 and #FFFFFF.');
    expect(Utils.getReverseColor.bind(Utils, 'aBcDeF')).to.throw('`color` must be a positive integer with hex format between #000000 and #FFFFFF.');
    expect(Utils.getReverseColor.bind(Utils, '#aBcDeF')).to.throw('`color` must be a positive integer with hex format between #000000 and #FFFFFF.');
    expect(Utils.getReverseColor.bind(Utils, [1, 2])).to.throw('`color` must be a positive integer with hex format between #000000 and #FFFFFF.');
    expect(Utils.getReverseColor.bind(Utils, {number: 1})).to.throw('`color` must be a positive integer with hex format between #000000 and #FFFFFF.');
    expect(Utils.getReverseColor.bind(Utils, null)).to.throw('`color` must be a positive integer with hex format between #000000 and #FFFFFF.');
    expect(Utils.getReverseColor.bind(Utils, undefined)).to.throw('`color` must be a positive integer with hex format between #000000 and #FFFFFF.');

    // Expected results
    expect(Utils.getReverseColor('1')).to.equal('#FFFFFE');
    expect(Utils.getReverseColor('11111')).to.equal('#FEEEEE');
    expect(Utils.getReverseColor('#11111')).to.equal('#FEEEEE');
    expect(Utils.getReverseColor('#abcdef')).to.equal('#543210');
    expect(Utils.getReverseColor('#FEDCBA')).to.equal('#012345');
  });

  it('generateNaturalSequence()', () => {
    // Invalid arguments
    expect(Utils.generateNaturalSequence.bind(Utils, 0)).to.throw('`count` must be a positive integer.');
    expect(Utils.generateNaturalSequence.bind(Utils, -1)).to.throw('`count` must be a positive integer.');
    expect(Utils.generateNaturalSequence.bind(Utils, 3.5)).to.throw('`count` must be a positive integer.');
    expect(Utils.generateNaturalSequence.bind(Utils, '1')).to.throw('`count` must be a positive integer.');
    expect(Utils.generateNaturalSequence.bind(Utils, [1, 2])).to.throw('`count` must be a positive integer.');
    expect(Utils.generateNaturalSequence.bind(Utils, {number: 1})).to.throw('`count` must be a positive integer.');
    expect(Utils.generateNaturalSequence.bind(Utils, null)).to.throw('`count` must be a positive integer.');
    expect(Utils.generateNaturalSequence.bind(Utils, undefined)).to.throw('`count` must be a positive integer.');

    // Expected results
    expect(Utils.generateNaturalSequence(1)).to.eql([1]);
    expect(Utils.generateNaturalSequence(10)).to.eql([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
  });

  it('generateNaturalMatrix()', () => {
    // Invalid arguments
    expect(Utils.generateNaturalMatrix.bind(Utils, 0)).to.throw('`rowCount` must be a positive integer.');
    expect(Utils.generateNaturalMatrix.bind(Utils, -1)).to.throw('`rowCount` must be a positive integer.');
    expect(Utils.generateNaturalMatrix.bind(Utils, 3.5)).to.throw('`rowCount` must be a positive integer.');
    expect(Utils.generateNaturalMatrix.bind(Utils, '1')).to.throw('`rowCount` must be a positive integer.');
    expect(Utils.generateNaturalMatrix.bind(Utils, [1, 2])).to.throw('`rowCount` must be a positive integer.');
    expect(Utils.generateNaturalMatrix.bind(Utils, {number: 1})).to.throw('`rowCount` must be a positive integer.');
    expect(Utils.generateNaturalMatrix.bind(Utils, null)).to.throw('`rowCount` must be a positive integer.');
    expect(Utils.generateNaturalMatrix.bind(Utils, undefined)).to.throw('`rowCount` must be a positive integer.');

    // Expected results
    expect(Utils.generateNaturalMatrix(1)).to.have.lengthOf(1);
    expect(Utils.generateNaturalMatrix(1)[0]).to.have.lengthOf(1);
    expect(Utils.generateNaturalMatrix(1)).to.eql([
                                                    [1]
                                                  ]); // deep equal
    let numbers4x4 = Utils.generateNaturalMatrix(4);
    expect(numbers4x4).to.have.lengthOf(4);
    numbers4x4.forEach(row => {
      expect(row).to.have.lengthOf(4);
    });
    expect(numbers4x4).to.eql(
      [
        [1, 2, 3, 4],
        [5, 6, 7, 8],
        [9, 10, 11, 12],
        [13, 14, 15, 16]
      ]
    ); // deep equal
  });

  it('matrixToArray()', () => {
    // Invalid arguments
    expect(Utils.matrixToArray.bind(Utils, 0)).to.throw('`matrix` must be a two-dimension array.\nYour `matrix` is: 0');
    expect(Utils.matrixToArray.bind(Utils, -1)).to.throw('`matrix` must be a two-dimension array.\nYour `matrix` is: -1');
    expect(Utils.matrixToArray.bind(Utils, 3.5)).to.throw('`matrix` must be a two-dimension array.\nYour `matrix` is: 3.5');
    expect(Utils.matrixToArray.bind(Utils, '1')).to.throw('`matrix` must be a two-dimension array.\nYour `matrix` is: 1');
    expect(Utils.matrixToArray.bind(Utils, {number: 1})).to.throw('`matrix` must be a two-dimension array.\nYour `matrix` is: [object Object]');
    expect(Utils.matrixToArray.bind(Utils, null)).to.throw('`matrix` must be a two-dimension array.\nYour `matrix` is: null');
    expect(Utils.matrixToArray.bind(Utils, undefined)).to.throw('`matrix` must be a two-dimension array.\nYour `matrix` is: undefined');
    expect(Utils.matrixToArray.bind(Utils, [1, 2, 3])).to.throw('`matrix` must be a two-dimension array.\nYour `matrix` is: 1,2,3');

    // Expected results
    expect(Utils.matrixToArray(
      [
        [1, 2, 3, 4],
        [5, 6, 7, 8],
        [9, 10, 11, 12],
        [13, 14, 15, 16]
      ]
    )).to.eql(Utils.generateNaturalSequence(16)); // deep equal
  });

  it('arrayToMatrix()', () => {
    // Invalid arguments
    expect(Utils.arrayToMatrix.bind(Utils, 0)).to.throw('`array` must be a one-dimension array.\nYour `array` is: 0');
    expect(Utils.arrayToMatrix.bind(Utils, -1)).to.throw('`array` must be a one-dimension array.\nYour `array` is: -1');
    expect(Utils.arrayToMatrix.bind(Utils, 3.5)).to.throw('`array` must be a one-dimension array.\nYour `array` is: 3.5');
    expect(Utils.arrayToMatrix.bind(Utils, '1')).to.throw('`array` must be a one-dimension array.\nYour `array` is: 1');
    expect(Utils.arrayToMatrix.bind(Utils, {number: 1})).to.throw('`array` must be a one-dimension array.\nYour `array` is: [object Object]');
    expect(Utils.arrayToMatrix.bind(Utils, null)).to.throw('`array` must be a one-dimension array.\nYour `array` is: null');
    expect(Utils.arrayToMatrix.bind(Utils, undefined)).to.throw('`array` must be a one-dimension array.\nYour `array` is: undefined');
    expect(Utils.arrayToMatrix.bind(Utils, [1, 2, 3])).to.throw('`array` must be a one-dimension array of n^2 members.\nYour `array` is: 1,2,3\nThere are 3 members in total.');
    expect(Utils.arrayToMatrix.bind(Utils, [1, 2, 3, 4], null, 2)).to.throw('`count` and `rowCount` must be integers, and `count` must be a square of `rowCount`.');
    expect(Utils.arrayToMatrix.bind(Utils, [1, 2, 3, 4], 3)).to.throw('`count` and `rowCount` must be integers, and `count` must be a square of `rowCount`.');
    expect(Utils.arrayToMatrix.bind(Utils, [1, 2, 3, 4], 3, 2)).to.throw('`count` and `rowCount` must be integers, and `count` must be a square of `rowCount`.');
    expect(Utils.arrayToMatrix.bind(Utils, [1, 2, 3, 4], 9, 3)).to.throw('`array` must be a one-dimension array of 9 members.\nYour `array` is: 1,2,3,4');

    // Expected results
    expect(Utils.arrayToMatrix([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16])).to.eql(
      [
        [1, 2, 3, 4],
        [5, 6, 7, 8],
        [9, 10, 11, 12],
        [13, 14, 15, 16]
      ]
    ); // deep equal
    expect(Utils.arrayToMatrix([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16], 16, 4)).to.eql(
      [
        [1, 2, 3, 4],
        [5, 6, 7, 8],
        [9, 10, 11, 12],
        [13, 14, 15, 16]
      ]
    ); // deep equal
  });
});

karma.conf.js
// Karma configuration
var webpack = require('webpack');

module.exports = function(config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',


    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['mocha'],


    // list of files / patterns to load in the browser
    files: [
      'test/js/**/*.spec.js'
    ],


    // list of files / patterns to exclude
    exclude: [
    ],


    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
      'test/js/**/*.spec.js': ['webpack']
    },


    webpack: {
      mode: 'development',
      //entry: ['babel-polyfill', __dirname + '/src/js/index.js'],
      resolve: {
        modules: [__dirname, 'node_modules'], // must tell test suite where src is
        extensions: ['.js']
      },
      module: {
        rules: [{
          exclude: [/node_modules/, __dirname + "/src/js/lib"],
          loader: 'babel',
          query: {
            compact: false,
            presets: ['es2015']
          }
        }]
      },
      resolveLoader: {
        moduleExtensions: ['-loader']
      },
      plugins: [
        new webpack.ProvidePlugin({
          _: 'lodash',
          assert: 'assert'
        })
      ]
    },


    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['progress'],


    // web server port
    port: 9876,


    // enable / disable colors in the output (reporters and logs)
    colors: true,


    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,


    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,


    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    //browsers: ['Chrome', 'ChromeCanary', 'PhantomJS'],


    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: false,

    // Concurrency level
    // how many browser should be started simultaneous
    concurrency: Infinity
  })
};

src/js/include.js
/**
 * Wrap common libs
 */
export default {
  /*
   * NodeJS assert
   */
  assert: typeof assert === 'undefined' ? (condition, msg) => {
    if (!condition) {
      throw new Error(msg);
    }
  } : assert,

  /*
   * Lodash
   */
  _: typeof _ === 'undefined' ? {
    isInteger: Number.isInteger,
    isString: s => typeof s === 'string',
    isArray: Array.isArray,
    parseInt: Number.parseInt,
    indexOf: (array, item) => array.indexOf(item),
    includes: (array, item) => array.indexOf(item) >= 0,
    each: (obj, fn) => {
      if (Array.isArray(obj)) {
        obj.forEach(fn);
      } else if (Object.values) {
        Object.values(obj).forEach(fn);
      }
      return obj;
    }, // no return value
    map: (array, fn) => array.map(fn),
    concat: (array1, array2) => array1.concat(array2),
    reverse: array => array.reverse(),
    reduce: (array, fn, accumulator) => array.reduce(fn),
    take: (array, n) => {
      if (Number.isInteger(n) && n > 0) {
        if (Array.isArray(array)) {
          return array.filter((item, index) => index <= n - 1);
        } else if (typeof array === 'string') {
          return array.substring(0, n).split('');
        } else {
          return [];
        }
      } else {
        if (Array.isArray(array) && array.length > 0) {
          return array[0];
        } else if (typeof array === 'string' && array.length > 0) {
          return [array.charAt(0)];
        } else {
          return [];
        }
      }
    },
    takeRight: (array, n) => {
      if (Number.isInteger(n) && n > 0) {
        if (Array.isArray(array)) {
          return array.filter((item, index) => index >= array.length - n && index < array.length);
        } else if (typeof array === 'string') {
          return array.substring(array.length - n).split('');
        } else {
          return [];
        }
      } else {
        if (Array.isArray(array) && array.length > 0) {
          return array[array.length - 1];
        } else if (typeof array === 'string' && array.length > 0) {
          return [array.charAt(array.length - 1)];
        } else {
          return [];
        }
      }
    },
    cloneDeep: obj => JSON.parse(JSON.stringify(obj))
  } : _
};

src/js/utils.js
import libs from './include';
const assert = libs.assert;
const _ = libs._;

export default {
  /**
   * Generate an array of colors from dark to light.
   *
   * @param count
   * @return {Array}
   */
  generateGradualColors: count => {
    assert(_.isInteger(count) && count > 0, '`count` must be a positive integer.');

    let colors = [];
    let max = 256;// FF+1
    let pace = max / count;
    let counter = 0;

    for (; counter / pace < count; counter += pace) {
      let color = counter.toString(16).toUpperCase() + '0000';
      if (color.length < 6) color = '0' + color;
      colors.push('#' + color);
    }

    return colors;
  },

  /**
   * Generate a matrix of colors from dark to light.
   *
   * @param rowCount
   * @return {Array} a square matrix with rowCount*rowCount color codes.
   */
  generateGradualColorMatrix: rowCount => {
    assert(_.isInteger(rowCount) && rowCount > 0, '`rowCount` must be a positive integer.');

    let colors = [];
    let max = 256;// FF+1
    let pace = max / rowCount;
    let rowCounter = 0;

    for (let i = 0; i < rowCount; i++) {
      let colCounter = 0;
      let row = [];
      for (let j = 0; j < rowCount; j++) {
        let red = rowCounter.toString(16).toUpperCase();
        if (red.length < 2) red = '0' + red;
        let green = colCounter.toString(16).toUpperCase();
        if (green.length < 2) green = '0' + green;
        row.push('#' + red + green + '00');
        colCounter += pace;
      }
      colors.push(row);
      rowCounter += pace;
    }

    return colors;
  },

  /**
   * @param color a string of hex number that represents R/G/B. e.g., "FF0000".
   * @return {string} the string value of the reverse color. e.g., "00FFFF".
   */
  getReverseColor: color => {
    assert(_.isString(color) && /^#?([0-9a-f]{1,6}|[0-9A-F]{1,6})$/g.test(color),
      '`color` must be a positive integer with hex format between #000000 and #FFFFFF.');

    let value = _.parseInt(color.replace(/#/g, ''), 16);
    let result = '000000' + (0xFFFFFF - value).toString(16).toUpperCase();

    return '#' + result.substring(result.length - 6, result.length);
  },

  /**
   * Create natural numbers sequence [1, 2, ..., n]
   *
   * @param count the count of numbers.
   * @return {number[]} the natural sequence.
   */
  generateNaturalSequence: count => {
    assert(_.isInteger(count) && count > 0,
      '`count` must be a positive integer.');

    return Array.from(new Array(count), (value, index) => index + 1);
  },

  /**
   * Create a two-dimension array of natural numbers.
   *
   *  [
   *   [1, 2, ..., k],
   *   ...,
   *   [..., n]
   * ]
   *
   * n = k ^ 2
   *
   * @param rowCount the length (k) of the matrix.
   * @return {number[][]} the matrix of natural numbers.
   */
  generateNaturalMatrix: rowCount => {
    assert(_.isInteger(rowCount) && rowCount > 0,
      '`rowCount` must be a positive integer.');

    return Array.from(new Array(rowCount), (value, rowIndex) =>
                      Array.from(new Array(rowCount), (value, colIndex) =>
                                 rowIndex * rowCount + colIndex + 1));
  },

  /**
   * Convert a two-dimension matrix to a one-dimension array.
   *
   * @param matrix a two-dimension array.
   * @return {Array} the flatted matrix in one-dimension array.
   */
  matrixToArray: matrix => {
    assert(_.isArray(matrix) && _.reduce(matrix, (accumulator, row) => accumulator && _.isArray(row)),
      '`matrix` must be a two-dimension array.\n' +
      'Your `matrix` is: ' + matrix);

    return _.reduce(matrix, (accumulator, row) => accumulator.concat(row));
  },

  /**
   * Convert an array of n^2 elements to a square matrix.
   *
   * @param array must be an array of n^2 elements.
   * @param (optional) count the length of the original array.
   * @param (optional) rowCount the length of the row count of the original matrix.
   * @return {number[][] | *[][]} a two-dimension square matrix.
   */
  arrayToMatrix: (array, count, rowCount) => {
    assert(_.isArray(array) && array.length > 0,
      '`array` must be a one-dimension array.\n' +
      'Your `array` is: ' + array);
    let rCount = Math.sqrt(array.length);
    assert(_.isInteger(rCount),
      '`array` must be a one-dimension array of n^2 members.\n' +
      'Your `array` is: ' + array + '\n' +
      'There are ' + array.length + ' members in total.');

    if (count !== undefined || rowCount !== undefined) {
      assert(_.isInteger(count) && count > 0 &&
             _.isInteger(rowCount) && rowCount > 0 &&
             count === rowCount ** 2,
        '`count` and `rowCount` must be integers, and `count` must be a square of `rowCount`.');
      assert(array.length === count,
        '`array` must be a one-dimension array of ' + count + ' members.\nYour `array` is: ' + array);
    }

    return Array.from(new Array(rCount), (val, rowIndex) =>
                      Array.from(new Array(rCount), (val, colIndex) =>
                                 array[rowIndex * rCount + colIndex]));
  }
}